https and token authentication for theia-ide

dealfonso opened the pull request:
# https and token authentication for theia-ide

This image adds **token authentication** and **https** to theia, to make your coding sessions more secure. The security is added by means of [gen-http-proxy](https://www.npmjs.com/package/gen-http-proxy), which is a generic http proxy that enables https and token authentication.

## Usage

Run on https://localhost:10443 with the current directory as a workspace (creating a random token):

```console
docker run --init -it -p 10443:10443 -v "$(pwd):/home/project:cached" theiaide/theia-https:latest
generating key file for https
generating cert file for https
root INFO Theia app listening on http://localhost:3000.
redirecting to localhost:3000
access url: https://0.0.0.0:10443?token=0f34fd329f266caca309bd47f8f8cc6f
token: 0f34fd329f266caca309bd47f8f8cc6f
use cookies: true
expiration: 60
```

## https and token authentication for other theia images

It should be possible to use it to add https and token authentication to _virtually any_ theia docker image (provided that the image makes theia listen at port 3000).

**example**

If you want to add the security layer to other theia app from the `theiaide` dockerhub repository, you can select the app (e.g. theia-go):

```console
$ docker build . --build-arg app=theia-go -t theiaide/theia-go-https
```

marcdumais-work commented:
@dealfonso interesting idea, thanks for this contribution. We'll try to try/review it ASAP.

vince-fugnitto commented:
@dealfonso do you plan on building and testing it using CI?
I think it'd be better, so it's necessary to add it to the existing `.travis.yml` file.
You can see other images defined, for example:

https://github.com/theia-ide/theia-apps/blob/61b06daeeb1ad6747e30e2560155a386946ea66b/.travis.yml#L13-L32

vince-fugnitto commented on sec-theia-docker/README.md:
```diff
@@ -0,0 +1,59 @@
+# Security layer for theia-ide
+
+[theia-ide](https://hub.docker.com/r/theiaide/theia) is a powerful mechamism for remote development in your servers. The problem is that it has not authentication mechanism, so if anyone knows your server and port, your code will be exposed.
```
typo - `mechamism` should be `mechanism`.

vince-fugnitto replied:
typo - `the problem is that it has not authentication mechanism` should be `the problem is that it has no authentication mechanism` (**not** to **no**)

vince-fugnitto commented on sec-theia-docker/README.md:
````diff
@@ -0,0 +1,59 @@
+If you want to create a theia-ide server for the current folder (with https and token authentication), you can issue the next command:
+
+```
+$ docker run -it --init -p 10443:10443 -v $(pwd):/home/project calfonso/sec-theia-ide
````
I don't think this command should reference your user or folder `calfonso`.

vince-fugnitto commented on sec-theia-docker/README.md:
````diff
@@ -0,0 +1,59 @@
+An example of changing the default port in which listens the secure theia ide is the next:
+
+```
+$ docker run -u $(id -u) --rm -it -p 10080:10080 -e server=:10080 -e secure=0 -v $(pwd):/home/project calfonso/sec-theia-ide
````
Again, no reference to `calfonso`.

vince-fugnitto commented on sec-theia-docker/README.md:
````diff
@@ -0,0 +1,59 @@
+Other example:
+
+```
+$ docker run -u $(id -u) --rm -it -p 20000:20000 -e server=:20000 -e token=mysecretpassword -v $(pwd):/home/project calfonso/sec-theia-ide
````
Same as above.

vince-fugnitto commented on sec-theia-docker/theiahere:
```diff
@@ -0,0 +1,22 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# 
+docker run -u $(id -u) --rm -it -p 10443:10443 -v $(pwd):/home/project calfonso/sec-theia-ide
```
I believe this command should be updated.
It will not work in my case or in general.

vince-fugnitto commented on sec-theia-docker/README.md:
```diff
@@ -0,0 +1,59 @@
+
+[theia-ide](https://hub.docker.com/r/theiaide/theia) is a powerful mechamism for remote development in your servers. The problem is that it has not authentication mechanism, so if anyone knows your server and port, your code will be exposed.
+
+This image enables to add a security layer for the standard image of theia-ide. It adds token authentication and a https layer to make your coding sessions more secure. The security is added by means of [gen-http-proxy](https://www.npmjs.com/package/gen-http-proxy), which is a generic http proxy that enables https and token authentication.
```
typo - it should be `This image adds a security layer...`

vince-fugnitto commented:
@dealfonso can you please ping me when you are ready for a second round of reviews? :)
Can you also please make sure to squash commits (unless it makes sense to have separate commits).

dealfonso commented:
@vince-fugnitto I am trying to run the tests, but I cannot make them pass. When I try them in my local hosts they pass (i.e. using yarn test-app-theia-sec), but travis cannot find the server... can you take a look or tell me how to get the screenshots?

dealfonso commented:
@vince-fugnitto after a lot of commits to guess why tests did not pass, I have found that the version of chrome in travis is a little buggy. Even passing the option `acceptSslCerts`, it has a problem with self signed certificates that can be shown in [this link](https://bugs.chromium.org/p/chromium/issues/detail?id=721739). Now the tests are passing, as expected.

vince-fugnitto commented:
@dealfonso 
Thanks for the ping! Do you mind updating the PR description to clearly describe how it can be tested? For example, adding security to the `theia-docker` image? It isn't clear at the moment nor is the 'readme' unfortunately.

marcdumais-work commented:
One general comment: would it be clearer what that image does, at a glimpse, if renamed from `theia-sec-docker` to e.g. `theia-https-docker`?

dealfonso commented:
> One general comment: would it be clearer what that image does, at a glimpse, if renamed from `theia-sec-docker` to e.g. `theia-https-docker`?

I agree... I am changing a pair of things and I will commit again

dealfonso commented:
@vince-fugnitto I updated the PR description, along to the readme file. I tried to make it more simple.
@marcdumais-work I renamed theia-sec-docker to theia-https-docker

Moreover I updated the dockerfile to build upon other theia-apps and base images (other than alpine).

dealfonso commented:
@vince-fugnitto have you checked this?

vince-fugnitto commented:
> @vince-fugnitto have you checked this?

I'm very sorry, I must have missed the notification.
I will perform another round of review promptly :)

vince-fugnitto commented on tests/wdio.theia-https.conf.js:
```diff
@@ -0,0 +1,31 @@
+    throw new Error(`${cliPortKey} expects a number as following argument`);
+}
+
+/*  The tests call the url / so we are not able to use the token. 
```
Do we still want to keep this comment or is it leftover?

dealfonso replied:
I'd prefer to keep it, in order to include a future check for using tokens.

vince-fugnitto replied:
No problem it's up to you :)

vince-fugnitto commented:
I manually restarted the build since it was failing prematurely.

dealfonso commented:
> I manually restarted the build since it was failing prematurely.

hi @vince-fugnitto, it seems that travis had problems yesterday and I had to restart the checks several times. But at the end the tests passed.

Could you please test it again?

dealfonso commented:
> I attempted the following:
> 
> _Build_:
> 
> ```shell
> docker build . -t theiaide/theia-https:latest
> ```
> 
> _Run_:
> 
> ```shell
> docker run --init -it -p 10443:10443 -v "$(pwd):/home/project:cached" theiaide/theia-https:latest
> ```
> 
> _Output_:
> 
> ```
> generating key file for https
> generating cert file for https
> could not spawn theia
> ```
> 
> Am I doing something wrong while testing?

The problem was the new version of node, that checks for package.json in the current folder. Now it has been corrected in the Docker file.

vince-fugnitto commented on build_container.sh:
```diff
@@ -5,6 +5,11 @@ set -e
 IMAGE_NAME=$2
 NODEVERSION=$3
+PORT=${4:-3000}
+shift
```
@dealfonso do you mind explaining what you are attempting to perform here or clean up?
I think in the long run it might be good to document the lines so it'd be easier to maintain and understand in the future.

vince-fugnitto replied:
I'm fine with the logic of it :)
Just wanted to see if we could add some documentation so it'd be easier to follow for future references.

dealfonso replied:
@vince-fugnitto do you need any extra work from my side?

vince-fugnitto replied:
@dealfonso only if you could provide some comments or documentation in the `build_container.sh` regarding the `shift`? It is not critical, but it will help future developers and also external contributors in better understanding what is going on and why it is needed. Then I think the PR is ready :)

marcdumais-work replied:
@dealfonso I think your excellent explanation above would work nicely if you could add it to the script.

@vince-fugnitto WDYT?

dealfonso replied:
Hi @marcdumais-work and @vince-fugnitto,

I have just included the comments in the script.

vince-fugnitto commented:
> > I attempted the following:
> > _Build_:
> > ```shell
> > docker build . -t theiaide/theia-https:latest
> > ```
> > 
> > 
> > _Run_:
> > ```shell
> > docker run --init -it -p 10443:10443 -v "$(pwd):/home/project:cached" theiaide/theia-https:latest
> > ```
> > 
> > 
> > _Output_:
> > ```
> > generating key file for https
> > generating cert file for https
> > could not spawn theia
> > ```
> > 
> > 
> > Am I doing something wrong while testing?
> 
> The problem was the new version of node, that checks for package.json in the current folder. Now it has been corrected in the Docker file.

I was able to build successfully :)

dealfonso commented:
> > > I attempted the following:
> > > _Build_:
> > > ```shell
> > > docker build . -t theiaide/theia-https:latest
> > > ```
> > > 
> > > 
> > > _Run_:
> > > ```shell
> > > docker run --init -it -p 10443:10443 -v "$(pwd):/home/project:cached" theiaide/theia-https:latest
> > > ```
> > > 
> > > 
> > > _Output_:
> > > ```
> > > generating key file for https
> > > generating cert file for https
> > > could not spawn theia
> > > ```
> > > 
> > > 
> > > Am I doing something wrong while testing?
> > 
> > 
> > The problem was the new version of node, that checks for package.json in the current folder. Now it has been corrected in the Docker file.
> 
> I was able to build successfully :)

Nice! so please make me know if you need additional feedback from my side.

vince-fugnitto commented:
> > > > I attempted the following:
> > > > _Build_:
> > > > ```shell
> > > > docker build . -t theiaide/theia-https:latest
> > > > ```
> > > > 
> > > > 
> > > > _Run_:
> > > > ```shell
> > > > docker run --init -it -p 10443:10443 -v "$(pwd):/home/project:cached" theiaide/theia-https:latest
> > > > ```
> > > > 
> > > > 
> > > > _Output_:
> > > > ```
> > > > generating key file for https
> > > > generating cert file for https
> > > > could not spawn theia
> > > > ```
> > > > 
> > > > 
> > > > Am I doing something wrong while testing?
> > > 
> > > 
> > > The problem was the new version of node, that checks for package.json in the current folder. Now it has been corrected in the Docker file.
> > 
> > 
> > I was able to build successfully :)
> 
> Nice! so please make me know if you need additional feedback from my side.

I think what's left is just in general to document this part, so it's easier to understand in the future
https://github.com/theia-ide/theia-apps/pull/243#discussion_r346281710. Then I think we'd be happy to merge the feature :)

marcdumais-work commented:
Very nice - I like how informative the README is, for this new `https` example app. Only a minor review comment left to address and it looks good to merge.

marcdumais-work commented on theia-https-docker/README.md:
````diff
@@ -0,0 +1,88 @@
+$ docker build . --build-arg version=next -t theiaide/theia-https:next
+```
+
+If you want to add the security layer to other theia app from the `theiaide` dockerhub repository, you can select the app (e.g. theia-go):
````
nice touch!

marcdumais-work commented:
I tested following instructions in the README and encountered no issue. Since the last review comment from @vince-fugnitto has been answered, I will now merge.

vince-fugnitto commented:
@marcdumais-work the CI was failing on the latest push :( I did not get a chance to comment on it before it was merged, but it looks like it failed for both the `theia-full` and `theia-https` images.

marcdumais-work commented:
> @marcdumais-work the CI was failing on the latest push :( I did not get a chance to comment on it before it was merged, but it looks like it failed for both the `theia-full` and `theia-https` images.

Yes, I saw. I think a small copy&paste error was introduced as part of the last update, duplicating some [code lines](https://github.com/theia-ide/theia-apps/pull/243/files#diff-61b99c5b8f88aa847fdaa60486fa1fabR8-R12). 

No biggie - I'm working on a quick PR: https://github.com/theia-ide/theia-apps/pull/273
